### PR TITLE
feat: Added `dnsResolver` Module

### PR DESCRIPTION
**Pull request by ChrisSidebotham**
## Description

Adding new module for `Microsoft.Network/dnsResolver` & extension resources in comparison to CARML current functionality.

| Pipeline | 
|----------|
| [avm.res.network.dns-resolver](https://github.com/ChrisSidebotham/bicep-registry-modules/actions/workflows/avm.res.network.dns-resolver.yml)|

### CODE REVIEW

File: avm/res/network/dnsresolver/README.md [hunk 1]
```diff
@@ -0,0 +1,527 @@
...
+| :-- | :-- |
+| `Microsoft.Authorization/locks` | [2020-05-01](https://learn.microsoft.com/en-us/azure/templates/Microsoft.Authorization/2020-05-01/locks) |
+| `Microsoft.Authorization/roleAssignments` | [2022-04-01](https://learn.microsoft.com/en-us/azure/templates/Microsoft.Authorization/2022-04-01/roleAssignments) |
+| `Microsoft.Network/dnsResolvers` | [2022-07-01](https://learn.microsoft.com/en-us/azure/templates/Microsoft.Network/2022-07-01/dnsResolvers) |
+| `Microsoft.Network/dnsResolvers/inboundEndpoints` | [2022-07-01](https://learn.microsoft.com/en-us/azure/templates/Microsoft.Network/2022-07-01/dnsResolvers/inboundEndpoints) |
+| `Microsoft.Network/dnsResolvers/outboundEndpoints` | [2022-07-01](https://learn.microsoft.com/en-us/azure/templates/Microsoft.Network/2022-07-01/dnsResolvers/outboundEndpoints) |
+
+## Usage examples
+
+The following module usage examples are retrieved from the content of the files hosted in the module's `tests` folder.
```
Comment (AlexanderSehr) [inline avm/res/network/dnsresolver/README.md:25]: Please regenerate with latest script from upstream. Some small things like the above will change

File: avm/res/network/dnsresolver/inbound-endpoint/main.bicep [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  name: dnsResolverName
+}
+
+resource dnsResolver_inboundEndpoint 'Microsoft.Network/dnsResolvers/inboundEndpoints@2022-07-01' = {
+  parent: dnsResolver
+  name: name
+  location: location
+  tags: tags
+  properties: {
+    ipConfigurations: [
```
Comment (AlexanderSehr) [inline avm/res/network/dnsresolver/inbound-endpoint/main.bicep:37]: I guess this is a 1:1 port from CARML right? If not, I'd be wondering if it's possible to only configure a single ip config per endpoint

File: avm/res/network/dnsresolver/inbound-endpoint/main.bicep [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+param privateIpAllocationMethod string = 'Dynamic'
+
+@description('Required. The name of the inbound endpoint.')
+param name string
+
+resource dnsResolver 'Microsoft.Network/dnsResolvers@2022-07-01' existing = {
+  name: dnsResolverName
+}
+
+resource dnsResolver_inboundEndpoint 'Microsoft.Network/dnsResolvers/inboundEndpoints@2022-07-01' = {
```
Comment (AlexanderSehr) [inline avm/res/network/dnsresolver/inbound-endpoint/main.bicep:31]:
```suggestion
resource inboundEndpoint 'Microsoft.Network/dnsResolvers/inboundEndpoints@2022-07-01' = {
```

Please also update references & files. 
We used to only use the `_` for cases were in this case `dnsResolver` would be deployed an be the main resource. Here however it is the `inboundEndpoint` throwing the party

File: avm/res/network/dnsresolver/inbound-endpoint/main.bicep [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        privateIpAllocationMethod: privateIpAllocationMethod
+      }
+    ]
+  }
+}
+
+@description('The name of the resource.')
+output name string = dnsResolver_inboundEndpoint.name
+
+@description('The ID of the resource.')
```
Comment (AlexanderSehr) [inline avm/res/network/dnsresolver/inbound-endpoint/main.bicep:52]:
```suggestion
@description('The resource ID of the resource.')
```

File: avm/res/network/dnsresolver/main.bicep [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+param lock lockType
+
+@description('Optional. Array of role assignment objects that contain the \'roleDefinitionIdOrName\' and \'principalId\' to define RBAC role assignments on this resource. In the roleDefinitionIdOrName attribute, you can provide either the display name of the role definition, or its fully qualified ID in the following format: \'/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11\'.')
+param roleAssignments roleAssignmentType
+
+@description('Optional. Tags of the resource.')
+param tags object?
+
+@description('Required. ResourceId of the virtual network to attach the Private DNS Resolver to.')
+param virtualNetworkId string
```
Comment (AlexanderSehr) [inline avm/res/network/dnsresolver/main.bicep:22]:
```suggestion
param virtualNetworkResourceId string
```
requires references to be updated too

File: avm/res/network/dnsresolver/outbound-endpoint/main.bicep [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+param dnsResolverName string
+
+@description('Optional. Location for all resources.')
+param location string = resourceGroup().location
+
+@description('Optional. Tags of the resource.')
+param tags object?
+
+@description('Required. The subnet ID of the inbound endpoint.')
+param subnetId string
```
Comment (AlexanderSehr) [inline avm/res/network/dnsresolver/outbound-endpoint/main.bicep:16]:
```suggestion
@description('Required. The subnet resource ID of the inbound endpoint.')
param subnetResourceId string
```

Requires updating in references too

File: avm/res/network/dnsresolver/outbound-endpoint/main.bicep [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+@description('Required. The name of the inbound endpoint.')
+param name string
+
+resource dnsResolver 'Microsoft.Network/dnsResolvers@2022-07-01' existing = {
+  name: dnsResolverName
+}
+
+resource dnsResolver_outboundEndpoint 'Microsoft.Network/dnsResolvers/outboundEndpoints@2022-07-01' = {
```
Comment (AlexanderSehr) [inline avm/res/network/dnsresolver/outbound-endpoint/main.bicep:25]:
```suggestion
resource outboundEndpoint 'Microsoft.Network/dnsResolvers/outboundEndpoints@2022-07-01' = {
```
Same comment as for inbound

File: avm/res/network/dnsresolver/outbound-endpoint/main.bicep [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    subnet: {
+      id: subnetId
+    }
+  }
+}
+
+@description('The name of the resource.')
+output name string = dnsResolver_outboundEndpoint.name
+
+@description('The ID of the resource.')
```
Comment (AlexanderSehr) [inline avm/res/network/dnsresolver/outbound-endpoint/main.bicep:40]:
```suggestion
@description('The resource ID of the resource.')
```

File: avm/res/network/dnsresolver/tests/e2e/defaults/main.test.bicep [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}-${iteration}'
+  params: {
+    name: '${namePrefix}${serviceShort}001'
+    virtualNetworkId: nestedDependencies.outputs.virtualNetworkId
+    tags: {
+      'hidden-title': 'This is visible in the resource name'
+      Environment: 'Non-Prod'
+      Role: 'DeploymentValidation'
+    }
```
Comment (AlexanderSehr) [inline avm/res/network/dnsresolver/tests/e2e/defaults/main.test.bicep:62]:
```suggestion
```

The tags rule should automatically be filtered out

File: avm/res/network/dnsresolver/tests/e2e/waf-aligned/main.test.bicep [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+      name: 'myCustomLockName'
+      kind: 'CanNotDelete'
+    }
+    roleAssignments: [
+      {
+        roleDefinitionIdOrName: 'Reader'
+        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
+        principalType: 'ServicePrincipal'
+      }
+    ]
```
[the remark this answers is not in the document]
Comment (ChrisSidebotham) [reply avm/res/network/dnsresolver/tests/e2e/waf-aligned/main.test.bicep:65]: @AlexanderSehr - We should consider changing this in current modules too then

File: avm/res/network/dnsresolver/README.md (hunk 1 above)
Comment (ChrisSidebotham) [reply avm/res/network/dnsresolver/README.md]: I believe this to now be complete with the latest changes @AlexanderSehr

File: avm/res/network/dnsresolver/inbound-endpoint/main.bicep (hunk 2 above)
Comment (ChrisSidebotham) [reply avm/res/network/dnsresolver/inbound-endpoint/main.bicep:37]: I recalled this not being possible beforehand - Having tested it, this is also correct, if you supply multiple IP's for the single endpoint this results in a exception being thrown from `Microsoft.Azure.Networking.Dns.ManagedResolver.Frontend.Contracts.Exceptions.Service.FrontendBadRequestLegacyServiceException`. The permitted experience is multiple InboundEndpoints with single IP Configurations. I believe the Array format comes as a requirement from frontDoor configuration for `Microsoft.Azure.Networking`

File: avm/res/network/dnsresolver/tests/e2e/waf-aligned/main.test.bicep (hunk 10 above)
[the remark this answers is not in the document]
Comment (ChrisSidebotham) [reply avm/res/network/dnsresolver/tests/e2e/waf-aligned/main.test.bicep]: updated in current version

[the remark this answers is not in the document]
Comment (AlexanderSehr) [reply avm/res/network/dnsresolver/tests/e2e/waf-aligned/main.test.bicep]: > @AlexanderSehr - We should consider changing this in current modules too then

The CARML modules you mean? In those we don't have a WAF folder

File: avm/res/network/dnsresolver/main.bicep [hunk 11]
```diff
@@ -102,7 +102,7 @@ module dnsResolver_inboundEndpoints 'inbound-endpoint/main.bicep' = [for (inboun
     tags: inboundEndpoint.?tags ?? tags
     location: inboundEndpoint.?location ?? location
     dnsResolverName: dnsResolver.name
-    subnetId: inboundEndpoint.subnetId
+    subnetResourceId: inboundEndpoint.subnetId
```
Comment (AlexanderSehr) [inline avm/res/network/dnsresolver/main.bicep:105]: I guess the reference here should also be `subnetResourceId`

Comment (AlexanderSehr) [reply avm/res/network/dnsresolver/main.bicep:105]: Could you check in the module overall if it is consistent?

File: avm/res/network/dnsresolver/tests/e2e/defaults/main.test.bicep [hunk 12]
```diff
@@ -54,18 +54,14 @@ module testDeployment '../../../main.bicep' = [for iteration in [ 'init', 'idem'
   name: '${uniqueString(deployment().name, location)}-test-${serviceShort}-${iteration}'
   params: {
     name: '${namePrefix}${serviceShort}001'
-    virtualNetworkId: nestedDependencies.outputs.virtualNetworkId
-    tags: {
-      'hidden-title': 'This is visible in the resource name'
-      Environment: 'Non-Prod'
-      Role: 'DeploymentValidation'
-    }
+    virtualNetworkResourceId: nestedDependencies.outputs.virtualNetworkId
```
Comment (AlexanderSehr) [inline avm/res/network/dnsresolver/tests/e2e/defaults/main.test.bicep:57]: Should ideally also reference a `resourceId`. I believe it's a good practice to be explicit by example

Comment (AlexanderSehr) [reply avm/res/network/dnsresolver/tests/e2e/defaults/main.test.bicep:57]: The same applies to the other test files

File: avm/res/network/dnsresolver/main.bicep (hunk 11 above)
Comment (ChrisSidebotham) [reply avm/res/network/dnsresolver/main.bicep]: Checked all files and module and replaced `virtualNetworkId` with `virtualNetworkResourceId` and `subnetId` with `subnetResourceId`

File: avm/res/network/dnsresolver/tests/e2e/defaults/main.test.bicep (hunk 12 above)
Comment (ChrisSidebotham) [reply avm/res/network/dnsresolver/tests/e2e/defaults/main.test.bicep]: see previous comments :)

File: avm/res/network/dnsresolver/tests/e2e/waf-aligned/main.test.bicep (hunk 10 above)
[the remark this answers is not in the document]
Comment (ChrisSidebotham) [reply avm/res/network/dnsresolver/tests/e2e/waf-aligned/main.test.bicep]: @AlexanderSehr - AVM Module, only Keyvault in main doesn't include this? unless i am mistaken?

[the remark this answers is not in the document]
Comment (AlexanderSehr) [reply avm/res/network/dnsresolver/tests/e2e/waf-aligned/main.test.bicep]: I'm not sure I follow :D The AVM Key Vault module has a WAF-test and it does not contain role assignments.

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**